### PR TITLE
Partial stabilization of `once_cell`

### PR DESCRIPTION
**Pull request by tgross35**
This PR aims to stabilize a portion of the `once_cell` feature:

- `core::cell::OnceCell`
- `std::cell::OnceCell` (re-export of the above)
- `std::sync::OnceLock` 

This will leave `LazyCell` and `LazyLock` unstabilized, which have been moved to the `lazy_cell` feature flag.

Tracking issue: https://github.com/rust-lang/rust/issues/74465 (does not fully close, but it may make sense to move to a new issue)

Future steps for separate PRs:
- ~~Add `#[inline]` to many methods~~ #105651
- Update cranelift usage of the `once_cell` crate
- Update rust-analyzer usage of the `once_cell` crate
- Update error messages discussing once_cell

## To be stabilized API summary

```rust
// core::cell (in core/cell/once.rs)

pub struct OnceCell<T> { .. }

impl<T> OnceCell<T> {
    pub const fn new() -> OnceCell<T>;
    pub fn get(&self) -> Option<&T>;
    pub fn get_mut(&mut self) -> Option<&mut T>;
    pub fn set(&self, value: T) -> Result<(), T>;
    pub fn get_or_init<F>(&self, f: F) -> &T where F: FnOnce() -> T;
    pub fn into_inner(self) -> Option<T>;
    pub fn take(&mut self) -> Option<T>;
}

impl<T: Clone> Clone for OnceCell<T>;
impl<T: Debug> Debug for OnceCell<T>
impl<T> Default for OnceCell<T>;
impl<T> From<T> for OnceCell<T>;
impl<T: PartialEq> PartialEq for OnceCell<T>;
impl<T: Eq> Eq for OnceCell<T>;
```

```rust
// std::sync (in std/sync/once_lock.rs)

impl<T> OnceLock<T> {
    pub const fn new() -> OnceLock<T>;
    pub fn get(&self) -> Option<&T>;
    pub fn get_mut(&mut self) -> Option<&mut T>;
    pub fn set(&self, value: T) -> Result<(), T>;
    pub fn get_or_init<F>(&self, f: F) -> &T where F: FnOnce() -> T;
    pub fn into_inner(self) -> Option<T>;
    pub fn take(&mut self) -> Option<T>;
}

impl<T: Clone> Clone for OnceLock<T>;
impl<T: Debug> Debug for OnceLock<T>;
impl<T> Default for OnceLock<T>;
impl<#[may_dangle] T> Drop for OnceLock<T>;
impl<T> From<T> for OnceLock<T>;
impl<T: PartialEq> PartialEq for OnceLock<T>
impl<T: Eq> Eq for OnceLock<T>;
impl<T: RefUnwindSafe + UnwindSafe> RefUnwindSafe for OnceLock<T>;
unsafe impl<T: Send> Send for OnceLock<T>;
unsafe impl<T: Sync + Send> Sync for OnceLock<T>;
impl<T: UnwindSafe> UnwindSafe for OnceLock<T>;
```

No longer planned as part of this PR, and moved to the `rust_cell_try` feature gate:

```rust
impl<T> OnceCell<T> {
    pub fn get_or_try_init<F, E>(&self, f: F) -> Result<&T, E> where F: FnOnce() -> Result<T, E>;
}

impl<T> OnceLock<T> {
    pub fn get_or_try_init<F, E>(&self, f: F) -> Result<&T, E> where F: FnOnce() -> Result<T, E>;
}
```

I am new to this process so would appreciate mentorship wherever needed.

## Post-merge update

With this merge, `LazyCell` and `LazyLock` have been moved under the gate `lazy_cell` (see #109736) and the `get_or_try_init` functions are under `once_cell_try` (see #109737)

### CODE REVIEW

**Comment by tgross35** [comment]
This needs FCP still

@rustbot modify labels: +T-libs-api

**Comment by programmerjake** [comment]
looks like you're stabilizing `OnceLock`, not `LazyCell`...the top comment should be fixed

**Comment by tgross35** [comment]
Updated, thanks for the catch

**Comment by klensy** [comment]
If this PR about stabilization, why slapping `inline` all over methods?

**Comment by tgross35** [comment]
I just did not know if those were typically added at this time, since these functions are fairly good use cases. I have removed them, matklad suggested we can add them later

File: library/core/src/cell.rs
Comment (programmerjake) [inline library/core/src/cell.rs:15]:
```suggestion
//! way - they do not implement [`Sync`]. (If you need to do aliasing and mutation among multiple
```
stray close paren

Comment (tgross35) [reply library/core/src/cell.rs]: Got it in the last rebase 👍  thanks

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by tgross35** [comment]
Rebased to fix conflicts

@bors label -S-waiting-on-author

**Comment by tgross35** [comment]
Whoops wrong robot

@rustbot label -S-waiting-on-author

**Comment by chris-morgan** [comment]
I was curious why this isn’t covering LazyCell and LazyLock, and it wasn’t immediately obvious to me.

After skimming a bit, I found https://github.com/rust-lang/rust/issues/74465#issuecomment-1316602359, which suggested that it was https://github.com/rust-lang/rust/pull/103718 (since landed) and possibly some concern about overlooking some suboptimality in a large surface area. I’m not sure what to make of that, because to me it looks like the Lazy parts are quite small, and the interesting surface area is largely in the Once parts.

So I’d just like to ask, as a general outsider to the process: is there actually any good reason for not just landing it all at this time?

**Comment by tgross35** [comment]
@chris-morgan matklad could answer this better than me, but 
I think the biggest thing is just that LazyCell/LazyLock haven’t been discussed in depth since OnceCell/OnceLock have gotten most of the attention, as the more primitive type. That’s not to say that the Lazy API will change much, but I think it’s easier to have that discussion when there’s no pressure about delaying Once types, (which currently seem to be used a lot more, not counting lazy_static). 

After this (hopefully) merges I’m going to ask for #74465 to be closed in favor of a new issue where LazyCell/LazyLock can get some of the discussion they deserve. Hopefully there won’t be any serious concerns there and a stabilization PR will only be a version or two behind this one.

But, I’m also generally an outsider here and these are just observations.

**Comment by matklad** [comment]
From my perspective: no mater how you slice it, the change here is unusually big in terms of API surface, so, barring very good reasons, the default answer should be to stabilize as little as possible. 

There are a bunch of small reasons why we might want to let LazyLock simmer a bit more: 

- there are very different genres of functionality, OnceLock adds a new fundamental capapability, while LazyLock is a convenience shortcut. It's like `std::fs::File` and `std::fs::read_to_string`. 
- lazy lock relies on generic default hack `F = fn() -> T`. This has tradeoffs, which needs to be argued about. In contrast, for OnceLock I feel we narrowed down design space sufficiently than there's a single obvious solution. This is a general point: "a single, obvious solution" is a major criterion for inclusion of the API in the std. 
- I am a bit unnerved that only recently we noticed https://github.com/rust-lang/rust/pull/103718

**Comment by matklad** [comment]
Also, to make sure we are explicit here: we indeed considered various naming schemes for **method** and decided that `_or_init` is the best we can do, right? 

This was discussed previously at https://github.com/rust-lang/rust/issues/74465#issuecomment-726020743.

**Comment by tgross35** [comment]
I think your comment there still sums up the situation well. The only real followup was from nwn here https://github.com/rust-lang/rust/issues/74465#issuecomment-726321236, which got a few thumbs up. That comment suggested that maybe `get_or_init` -> `get_or_init_with` and `get_or_try_init` -> `try_get_or_init_with` might be more consistent names for something with a closure. I don't disagree with this, since it would also leave room for a new `get_or_init` that takes a value rather than a closure (unlikely use case, but possible).

Added verbosity for this particular example, plus a last minute change if it's changed at all are annoying - but better to make sure it's consistent now than say "doh" down the line. Perhaps good to defer to / check with T-libs

**Comment by joboet** [comment]
Sorry for missing FCP here, but I'm quite concerned about stabilising `get_or_try_init` as is, because that makes it impossible to make that function genetic over its output type. For instance, it would be nice to allow returning an `Option` from the closure, which could be made possible with an extension to the `Try` trait (something like an extra `ChangeOutput` trait that uses GATs to allow this). If `get_or_try_init` is stable, this is a breaking change since users might have explicitly set the `E` generic.

**Comment by tgross35** [comment]
Hey @joboet do you have an example of what you'd like to do, or what the bounds would look like? In most cases where you'd want to use a closure that returns an Option, you could just do

```rust
my_cell.get_or_try_init(|| fn_returning_option().ok_or("this will be a Err(&str)"))
```

but unsure if you have something more specific

**Comment by matklad** [comment]
Good catch @joboet, this does seem like a new unresolved question which we haven’t covered before. We indeed have a couple of precedents in std of using `Try` trait in similar APIs:

* https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.try_for_each
* https://doc.rust-lang.org/stable/std/primitive.array.html#method.try_map

cc @scottmcm, our Try expert, in case they have any advice wrt to stabilizing this.

My thinking:

* we can cut `try` method from the initial stabilization. It’s a core API (you can express it via other methods), but OnceLock is useful without it.
* From practical perspective, I think this being non-generic over try is actually beneficial: I don’t think it’s useful to use anything but result here, and Try tends to make signatures inscrutable (and type inference more picky).
* From consistency perspective, we should use Try here, as that’s what used elsewhere, and the “correct” solution for stdlib.

**Comment by scottmcm** [comment]
#107122 suggests that `get_or_try_init` would mean using `Residual`, which means that based on things like https://github.com/rust-lang/rust/issues/87053#issuecomment-1385708795 it probably shouldn't be stabilized as generic just yet.

Stabilizing just the non-`try` version for now (as was done in #94119) would probably be the easiest way forward.

**Comment by tgross35** [comment]
That seems reasonable enough, I will remove the `try_` functions. Great catch @joboet 

I guess that leaves the last concern here as the name of  `get_or_init`. I kind of would like to double down on changing it to anything `_with` (e.g. `get_or_init_with`). That precedent is really strong for closure takers: resize_with, intersperse_with, fill_with, try_with, then_with, or_insert_with, zip_with, and (especially relevant) the neighbor RefCell::replace_with.

But, as matklad mentioned on Zulip, that’s up to the final reviewer

**Comment by tgross35** [comment]
@m-ou-se as the assignee on this one, would you mind sharing your opinion about the name?

I proposed in #107184 to rename `get_or_init` to `get_or_init_with` to be consistent with other cell types and other functions that do take closures, as described in that issue. I still maintain that that (or something else `*_with`) would be a better name for `std` consistency, but I also do not want to block this useful feature on bikeshedding.

Just to be clear on Scott's point above - I removed the `get_or_try_init` functions from this stabilization PR, so this is not blocked on `Try` trait decisions.

**Comment by tgross35** [comment]
Forgot to update the labels after my last rebase

@rustbot ready

**Comment by m-ou-se** [comment]
Sorry for my late reply.

It looks like two things happened after the FCP completed:

- A proposal to rename `get_or_init` to `get_or_init_with`, for concistency.
- A suggestion to not yet stabilize the `try` methods, since those currently use the `Result` type rather than the `Try` trait.

Both seem reasonable to me. This does need a new review from the team though, so I'll restart the FCP once https://github.com/rust-lang/rust/pull/107184 is merged.

**Comment by tgross35** [comment]
With discussion at https://github.com/rust-lang/rust/pull/107184 being finished, this is ready for merge. The conclusion was not to change the existing name.

The only change from FCP time was to remove the fallible methods, which are accurately reflected in this issue's description. I don't think there is anything there that would require going through FCP again.

@rustbot label +finished-final-comment-period +disposition-merge +to-announce -needs-fcp

**Review by m-ou-se — APPROVED** [review]
Thanks for also improving the documentation!

**Comment by m-ou-se** [comment]
@bors r-

Oh actually, maybe it's better to stabilize this just as `"once_cell"` rather than `"once_cell_min"`, and instead give a new feature name to the few remaining unstable items. Then users or `#![feature(once_cell)]` will get a message that says it has been stabilized. What do you think?

**Comment by tgross35** [comment]
Also, the RFC should probably be merged to clean up that PR https://github.com/rust-lang/rfcs/pull/2788 (or closed)

**Comment by tgross35** [comment]
Alright, CI finally passed with the gate changes. Ready for another r

@rustbot label -A-query-system -A-translation

**Pull request merged**
